frontend/src/packages/editor/operationManager.tsx

```diff
@@ -30,4 +30,98 @@ export const slateToCmsOperation = (editorContent: BlockData, operation: BaseOpe
             noop: {}
         }
     }
+}
+
+
+/** 
+ * The semantics of Slate Operations
+ *      - Slate models operations somewhat weirdly, theres a few key types
+ *          - Insert/Remove text modifies the text field so the "path" isn't actually a complete path in the traditional OT sense 
+ *          - Set-node modifies a specific field
+ *          - Paths like [a, b, c] refer to indexes in either array elements or children fields
+*/
+
+// normalizePath extends the path in a slate operation to also include the field it is editing
+// for example: if we receive the slate operation {set-node underline = true of text-object} with the path [0, 0, 0]
+// the normalised path will be [0, 0, 0, 4] (assuming 4 is index 4 in the text object)
+const normalizePath = (editorContent: BlockData, op: NodeOperation | TextOperation): number[] => {
+    // quirkiness with the CMS text operation interface, the target is included in the path
+    if (op.type === "remove_text" || op.type === "insert_text") {
+        return op.path;
+    }
+
+    // resolve what type we're studying and fetch the field mappings for it
+    return []
+}
+
+const convertTextInsertionOp = (editorContent: BlockData, op: InsertTextOperation): CMSOperation => (
+    {
+        Path: op.path,
+        OperationType: "insert",
+        IsNoOp: false,
+        Operation: {
+            $type: "stringOperation",
+            stringOperation: {
+                rangeStart: op.offset,
+                rangeEnd: op.offset + op.text.length,
+                newValue: op.text
+            }
+        }
+    }
+);
+
+
+const convertTextRemovalOp = (editorContent: BlockData, op: RemoveTextOperation): CMSOperation => (
+    {
+        Path: op.path,
+        OperationType: "delete",
+        IsNoOp: false,
+        Operation: {
+            $type: "stringOperation",
+            stringOperation: {
+                rangeStart: op.offset - op.text.length,
+                rangeEnd: op.offset,
+                newValue: op.text,
+            }
+        }
+    }
+);
+
+
+const normalizeElementPath = (contentBlock: CMSEditorNode, op: NodeOperation | TextOperation): number[] => {
+    switch (true) {
+        case IsCustomElement(contentBlock): return normalizeCustomElementPath(contentBlock as CustomElement, op);
+        case IsCustomTextBlock(contentBlock): return normalizeCustomTextPath(contentBlock as CustomText, op);
+    }
+
+    return [];
+}
+
+const normalizeCustomElementPath = (contentBlock: CustomElement, op: NodeOperation | TextOperation): number[] => {
+    return [] 
+}
+
+const normalizeCustomTextPath = (contentBlock: CustomText, op: NodeOperation | TextOperation): number[] => {
+    // TODO: For Mae :P - So the issue here is that fields on javascript objects are unordered, this is unlike Go where (typically) the order in which
```

Comment:
this isn't quite correct, objects do maintain some order:

> The traversal order, as of modern ECMAScript specification, is well-defined and consistent across implementations. Within each component of the prototype chain, all non-negative integer keys (those that can be array indices) will be traversed first in ascending order by value, then other string keys in ascending chronological order of property creation.

([MDN for..in](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in#description), this also extends to object methods like `Object.keys()` and `Object.entries()`)

Reply:
Oh  thats convenient, so should it directly correspond to the the order declared in the TS type?

Reply:
> Oh thats convenient, so should it directly correspond to the the order declared in the TS type?

Wait ig not since it depends on order of declaration in the prototype so ig anyone can declare it in any order they want

Reply:
not necessarily as declared in the type, as those aren't available at runtime - if you send an json object from the backend and parse that from the frontend, it *should* maintain whatever order it was sent in (assuming there's no integer keys)

Reply:
Hmm issue is though is that slate doesn't actually seem to give us a guarantee on order (for starters since it doesn't surface the entire object that its update and only surfaces the field)

Reply:
gg rip that's doomed then